### PR TITLE
Download and display images from public Firebase bucket.

### PR DESCRIPTION
**Pull request by glennhartmann**
Sits on top of PR https://github.com/GoogleChromeLabs/llaminator/pull/107 and contributes to Issue https://github.com/GoogleChromeLabs/llaminator/issues/103.

I say "contributes to" and not "fixes" because although we display the images
on the page, we don't actually save them into our local storage  yet (ie, if you
go offline, they disappear).

### CODE REVIEW

File: src/storage.ts [hunk 1]
```diff
@@ -14,12 +14,16 @@
...
+import * as firebase from 'firebase/storage';
 import { openDB, DBSchema, IDBPDatabase } from 'idb';
 import { v4 as uuidv4 } from 'uuid';
-import { ref, uploadBytes, UploadMetadata, UploadResult } from 'firebase/storage';
+
 import { firebaseStorage } from './firebase';
 
 const kMainDBName = 'appDB';
+const firebaseDir = 'public';
+const remoteIdPrefix = 'R-';
```
Comment (beverloo) [inline src/storage.ts:25]: style nit: we should be consistent w/ constant naming, so either all k-prefix, or none. (kFirebaseDir, kRemoteIdPrefix)

File: src/storage.ts [hunk 2]
```diff
@@ -123,10 +127,44 @@ export class LlamaStorage {
   }
 
   async list(): Promise<FileRecord[]> {
-    return this.db.getAll('metadata');
+    const localMeta = this.db.getAll('metadata');
+    let firebaseList: firebase.ListResult | null = null;
+    if (firebaseStorage) {
+      firebaseList = await firebase.listAll(firebase.ref(firebaseStorage, firebaseDir));
```
Comment (beverloo) [inline src/storage.ts:133]: Is it the right thing to do to sync this often?

An alternative approach to data consistency, having a slightly weaker contract, might be to have the local state be leading, but do an occasional sync w/ Firebase to identify differences (e.g. once per launch in the background, or using Periodic Background Sync). That might be kinder to offline users as well, and those on lie-fi.

File: src/storage.ts [hunk 3]
```diff
@@ -123,10 +127,44 @@ export class LlamaStorage {
...
+
+    // TODO: this is potentially high latency - can we load asynchronously?
+    for (const res of firebaseList.items) {
+      const meta = await this.getRemoteFirebaseMetadata(res);
+      if (meta) {
+        results.push(meta);
+      }
+    }
+
+    const m: { [key: string]: boolean } = {};
```
Comment (beverloo) [inline src/storage.ts:149]: nit: use a [Set](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set)

File: src/storage.ts [hunk 4]
```diff
@@ -123,10 +127,44 @@ export class LlamaStorage {
...
+    return results.filter((v: FileRecord) => {
+      let id = v.id;
+      if (id.indexOf(remoteIdPrefix) === 0) {
```
Comment (beverloo) [inline src/storage.ts:152]: nit: here and elsewhere, `id.startsWith(remoteIdPrefix)` ([mdn](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/startsWith)) which reads a lot more intuitively.

**Review by beverloo — APPROVED** [review]
Happy to proceed with this, but my sense for both PRs is that we'd want Firebase support to be bolted as a service on top of local storage (i.e. an external that can modify it when synchronisation happens) rather than completely integrated with local storage.

File: src/storage.ts (hunk 1 above)
Comment (glennhartmann) [reply src/storage.ts]: Done.

File: src/storage.ts (hunk 2 above)
Comment (glennhartmann) [reply src/storage.ts]: > Is it the right thing to do to sync this often?

Probably not - I'm mostly just figuring out how to make Firebase work at all for now, and worrying about efficiency later. Hooray for tech debt.

> (e.g. once per launch in the background, or using Periodic Background Sync).

Never used Periodic Background Sync, but it does sound reasonable. https://github.com/GoogleChromeLabs/llaminator/issues/110

File: src/storage.ts (hunk 3 above)
Comment (glennhartmann) [reply src/storage.ts]: Done.

File: src/storage.ts (hunk 4 above)
Comment (glennhartmann) [reply src/storage.ts]: Done.

**Comment by glennhartmann** [comment]
> Happy to proceed with this, but my sense for both PRs is that we'd want Firebase support to be bolted as a service on top of local storage (i.e. an external that can modify it when synchronisation happens) rather than completely integrated with local storage.

Yeah, agreed - not really happy with keeping the Firebase stuff in the storage class long-term.

**Pull request merged**